effect/jvm/src/main/scala/scalaz/effect/MVarImpl.scala

```diff
@@ -0,0 +1,284 @@
+// Copyright (C) 2017 John A. De Goes. All rights reserved.
+package scalaz.effect
+
+import scalaz.data.Maybe
+import scalaz.data.Disjunction._
+
+import java.util.concurrent.atomic.AtomicReference
+
+import MVarInternal._
+
+private[effect] final class MVarImpl[A](threadPool: (=> Unit) => Unit, val state: AtomicReference[MVarState[A]]) extends MVar[A] {
+  final def peek: IO[Maybe[A]] = IO.sync {
+    state.get match {
+      case Surplus(head, _) => Maybe.just(head)
+      case Deficit(_, _) => Maybe.empty[A]
+    }
+  }
+
+  final def take: IO[A] = IO.async0 { taker =>
+    var loop  = true
+    var finish: Action[A] = null
+
+    while (loop) {
+      finish = null
+
+      val oldState = state.get
+
+      val newState = oldState match {
+        case Surplus(value0, putters) =>
+          if (putters.length == 0) {
+            finish = () => AsyncReturn.now(value0)
+
+            Empty
+          } else {
+            val (value, putter0) = putters(0)
+
+            finish = () => {
+              threadPool(putter0(SuccessUnit))
+
+              AsyncReturn.now(value0)
+            }
+
+            Surplus(value, putters.tail)
+          }
+        case Deficit(takers, readers) =>
+          finish = () => AsyncReturn.maybeLater(removeTaker(taker, _))
+
+          Deficit(takers :+ taker, readers)
+      }
+
+      if (state.compareAndSet(oldState, newState)) loop = false
+    }
```

Comment:
Scala does have `do while` loop. 😉 
Here and in several other places below `var loop` can be replaced with a `do while` loop.